Source/JavaScriptCore/heap/FreeListInlines.h

```diff
@@ -33,36 +33,49 @@ namespace JSC {
 template<typename Func>
 ALWAYS_INLINE HeapCell* FreeList::allocate(const Func& slowPath)
 {
-    unsigned remaining = m_remaining;
-    if (remaining) {
-        unsigned cellSize = m_cellSize;
-        remaining -= cellSize;
-        m_remaining = remaining;
-        return bitwise_cast<HeapCell*>(m_payloadEnd - remaining - cellSize);
+    unsigned cellSize = m_cellSize;
+    if (LIKELY(m_intervalStart < m_intervalEnd)) {
+        char* result = m_intervalStart;
+        m_intervalStart += cellSize;
+        return bitwise_cast<HeapCell*>(result);
     }
     
-    FreeCell* result = head();
-    if (UNLIKELY(!result))
+    FreeCell* cell = nextInterval();
+    if (UNLIKELY(isSentinel(cell)))
         return slowPath();
+
+    FreeCell::advance(m_secret, m_nextInterval, m_intervalStart, m_intervalEnd);
     
-    m_scrambledHead = result->scrambledNext;
+    // We assume we don't create any empty intervals; so there should be enough space remaining
+    // to allocate a cell.
+    char* result = m_intervalStart;
+    m_intervalStart += cellSize;
     return bitwise_cast<HeapCell*>(result);
 }
 
 template<typename Func>
 void FreeList::forEach(const Func& func) const
 {
-    if (m_remaining) {
-        for (unsigned remaining = m_remaining; remaining; remaining -= m_cellSize)
-            func(bitwise_cast<HeapCell*>(m_payloadEnd - remaining));
-    } else {
-        for (FreeCell* cell = head(); cell;) {
-            // We can use this to overwrite free objects before destroying the free list. So, we need
-            // to get next before proceeding further.
-            FreeCell* next = cell->next(m_secret);
-            func(bitwise_cast<HeapCell*>(cell));
-            cell = next;
-        }
+    FreeCell* cell = nextInterval();
+    char* intervalStart = m_intervalStart;
+    char* intervalEnd = m_intervalEnd;
+    ASSERT(intervalEnd - intervalStart < (ptrdiff_t)(16 * KB));
+
+    // Somewhat tricky control flow here...we need to allow either of these conditions because:
+    //  - We might reach here where the head is sentinel, but there is a non-empty interval we 
+    //    can still traverse.
+    //  - The interval might be empty, but the head isn't the sentinel so we should just pop 
+    //    the next one.
+    while (intervalStart < intervalEnd || !isSentinel(cell)) {
+        for (; intervalStart < intervalEnd; intervalStart += m_cellSize)
+            func(bitwise_cast<HeapCell*>(intervalStart));
+
+        // If we explore the whole interval and the cell is the sentinel value, though, we should
+        // immediately exit so we don't decode anything out of bounds.
+        if (isSentinel(cell))
+            break;
+
+        FreeCell::advance(m_secret, cell, intervalStart, intervalEnd);
     }
```

Comment:
Maybe, this is cleaner?
```
while (true) {
    for (; intervalStart < intervalEnd; intervalStart += m_cellSize)
        func(bitwise_cast<HeapCell*>(intervalStart));
            
    // If we explore the whole interval and the cell is the sentinel value, though, we should
    // immediately exit so we don't decode anything out of bounds.
    if (isSentinel(cell))
        break;

    FreeCell::advance(m_secret, cell, intervalStart, intervalEnd);
}
```

Reply:
Oh, yes I agree, I'll quickly make that change.